### PR TITLE
add mb_check_encoding with array value

### PR DESCRIPTION
**Pull request by jeremyFreeAgent**
I tried to address the https://github.com/symfony/polyfill/issues/422 issue.

First time on that project. I'm not sure about what I've done because I'm afraid we cannot do anything for PHP <7.2.

### CODE REVIEW

**Comment by alexandre-daubois** [comment]
Caution, [the doc](https://www.php.net/manual/en/function.mb-check-encoding.php) tells that values passed as array are validated recursively 🙂

**Comment by jeremyFreeAgent** [comment]
> Caution, [the doc](https://www.php.net/manual/en/function.mb-check-encoding.php) tells that values passed as array are validated recursively 🙂

Thanks!

**Comment by derrabus** [comment]
Let's kill PHP 7.1 then, shall we? #428

File: src/Mbstring/Mbstring.php [hunk 1]
```diff
@@ -413,6 +413,19 @@ public static function mb_check_encoding($var = null, $encoding = null)
             $encoding = self::$internalEncoding;
         }
 
+        if (is_array($var)) {
```
Comment (nicolas-grekas) [inline src/Mbstring/Mbstring.php:416]: let's use an early exit:
`if (!\is_array()) ... return;`

File: tests/Mbstring/MbstringTest.php [hunk 2]
```diff
@@ -466,6 +466,38 @@ public function testCheckEncoding()
         $this->assertTrue(mb_check_encoding("\xE9", 'Windows-1252'));
     }
 
+    /**
+     * @covers \Symfony\Polyfill\Mbstring\Mbstring::mb_check_encoding
+     */
+    public function testCheckEncodingWithArrayValue()
+    {
+        if (70200 <= \PHP_VERSION_ID) {
```
Comment (nicolas-grekas) [inline tests/Mbstring/MbstringTest.php:474]: should be replaced with `@requires PHP 7.2`

**Comment by stof** [comment]
For PHP 7.1, we need to emulate the PHP 7.1 behavior of mbstring, which is not supporting arrays (as we cannot make a project using PHP 7.1 with native mbstring support arrays by installing the polyfill as it would not replace the native functions).
That's totally possible to emulate (triggering the warning and returning `null`)

**Comment by jeremyFreeAgent** [comment]
Thank you @stof and @nicolas-grekas for the PHP < 7.2 behaviour.

File: src/Mbstring/Mbstring.php [hunk 3]
```diff
@@ -406,14 +406,34 @@ public static function mb_encoding_aliases($encoding)
 
     public static function mb_check_encoding($var = null, $encoding = null)
     {
+        if (PHP_VERSION_ID < 70200 && \is_array($var)) {
+            trigger_error('mb_check_encoding() expects parameter 1 to be string, array given');
```
Comment (nicolas-grekas) [inline src/Mbstring/Mbstring.php:410]: `, \E_USER_WARNING` as second arg (by default it's a notice)

File: tests/Mbstring/MbstringTest.php [hunk 4]
```diff
@@ -466,6 +466,53 @@ public function testCheckEncoding()
...
+    /**
+     * @covers \Symfony\Polyfill\Mbstring\Mbstring::mb_check_encoding
+     *
+     * @requires PHP < 7.2
+     */
+    public function testCheckEncodingWithArrayValueForPhpLessThan72()
+    {
+        $errorMessage = null;
+        set_error_handler(function ($errno, $errstr, $errfile, $errline) use (&$errorMessage) { $errorMessage = $errstr; });
+        mb_check_encoding(['aςσb'], 'UTF8');
```
Comment (nicolas-grekas) [inline tests/Mbstring/MbstringTest.php:510]: this->assertNull

File: tests/Mbstring/MbstringTest.php [hunk 5]
```diff
@@ -466,6 +466,53 @@ public function testCheckEncoding()
...
+
+    /**
+     * @covers \Symfony\Polyfill\Mbstring\Mbstring::mb_check_encoding
+     *
+     * @requires PHP < 7.2
+     */
+    public function testCheckEncodingWithArrayValueForPhpLessThan72()
+    {
+        $errorMessage = null;
+        set_error_handler(function ($errno, $errstr, $errfile, $errline) use (&$errorMessage) { $errorMessage = $errstr; });
```
Comment (nicolas-grekas) [inline tests/Mbstring/MbstringTest.php:509]: `set_error_handler(function ($type, $msg, $file, $line) use (&$errorMessage) { $errorMessage = \E_USER_WARNING === $type ? $msg : null; });

File: src/Mbstring/Mbstring.php (hunk 3 above)
Comment (jeremyFreeAgent) [reply src/Mbstring/Mbstring.php]: Thanks, I saw it and fixed it!

File: tests/Mbstring/MbstringTest.php (hunk 4 above)
Comment (jeremyFreeAgent) [reply tests/Mbstring/MbstringTest.php]: Good catch!

File: tests/Mbstring/MbstringTest.php (hunk 5 above)
Comment (jeremyFreeAgent) [reply tests/Mbstring/MbstringTest.php]: Thanks

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @jeremyFreeAgent.

**Pull request merged**